contracts/marketpools.js

```diff
@@ -0,0 +1,346 @@
+    || !api.assert(baseQuantity && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
+    || !api.assert(quoteQuantity && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
+    || !await validateTokenPair(tokenPair)) return;
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const baseToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: baseSymbol });
+  const quoteToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: quoteSymbol });
+  if (!api.assert(api.BigNumber(baseQuantity).dp() <= baseToken.precision, 'baseQuantity precision mismatch')
+    || !api.assert(api.BigNumber(quoteQuantity).dp() <= quoteToken.precision, 'quoteQuantity precision mismatch')) return;
+
+  const pool = await api.db.findOne('pools', { tokenPair });
+  if (api.assert(pool, 'no existing pool for tokenPair')) {
+    if (api.BigNumber(pool.baseQuantity).eq(0) && api.BigNumber(pool.quoteQuantity).eq(0)
+      && await validateOracle(pool, api.BigNumber(quoteQuantity).dividedBy(baseQuantity)) === false) return;
+    if (api.BigNumber(pool.baseQuantity).gt(0) && api.BigNumber(pool.quoteQuantity).gt(0)
+      && !validateLiquiditySwap(pool, baseQuantity, quoteQuantity)) return;
+
+    const senderBase = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: baseSymbol });
+    const senderQuote = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: quoteSymbol });
+    const senderFunded = senderBase && senderQuote && api.BigNumber(senderBase.balance).gte(baseQuantity) && api.BigNumber(senderQuote.balance).gte(quoteQuantity);
+    if (!api.assert(senderFunded, 'insufficient token balance')) return;
+
+    // update liquidity position
+    const lp = await api.db.findOne('liquidityPositions', { account: api.sender, tokenPair });
+    if (lp) {
+      lp.baseQuantity = api.BigNumber(lp.baseQuantity).plus(baseQuantity);
+      lp.quoteQuantity = api.BigNumber(lp.quoteQuantity).plus(quoteQuantity);
+      await api.db.update('liquidityPositions', lp);
+    } else {
+      const newlp = {
+        account: api.sender,
+        tokenPair,
+        baseQuantity,
+        quoteQuantity,
+      };
+      await api.db.insert('liquidityPositions', newlp);
+    }
+
+    // deposit requested tokens to contract
+    const baseRes = await api.executeSmartContract('tokens', 'transferToContract', { symbol: baseSymbol, quantity: baseQuantity, to: 'marketpools' });
+    const quoteRes = await api.executeSmartContract('tokens', 'transferToContract', { symbol: quoteSymbol, quantity: quoteQuantity, to: 'marketpools' });
+    if (!api.assert(baseRes.errors === undefined && quoteRes.errors === undefined, 'deposit transfer errors')) return;
+    await updatePoolStats(pool, baseQuantity, quoteQuantity, false);
+    api.emit('addLiquidity', { memo: `Add ${baseSymbol} and ${quoteSymbol}` });
```

Comment:
this deviates from other such events. Can we change this to just  

{ baseSymbol, quoteSymbol }

Reply:
That should work. The reason I went with a readable string was figured this would go in a column on the frontends.

Reply:
I like eonwarped's suggestion here, front-ends can interpret and pretty-print it as they wish.